File: csharp/BCUR/BCUR.Tests/URCodableTests.cs
using BlockchainCommons.DCbor;

namespace BlockchainCommons.BCUR.Tests;

/// <summary>
/// A test type implementing URCodable, matching the Rust test in ur_codable.rs.
/// </summary>
internal sealed class TestLeaf : ICborTagged, IURCodable
{
    public string S { get; }

    public TestLeaf(string s) => S = s;

    public static IReadOnlyList<Tag> CborTags => [new Tag(24, "leaf")];

    public Cbor UntaggedCbor() => Cbor.FromString(S);

    public Cbor TaggedCbor() => Cbor.ToTaggedValue(CborTags[0], UntaggedCbor());

    public Cbor ToCbor() => TaggedCbor();

    public static TestLeaf FromUntaggedCbor(Cbor cbor)
    {
        return new TestLeaf(cbor.TryIntoText());
    }

    public static TestLeaf FromTaggedCbor(Cbor cbor)
    {
        var item = cbor.TryIntoExpectedTaggedValue(CborTags[0]);
        return FromUntaggedCbor(item);
    }
}

public class URCodableTests
{
    [Fact]
    public void UrCodableRoundTrip()
    {
        var test = new TestLeaf("test");
        var ur = test.ToUR();
        var urString = ur.ToUrString();
        Assert.Equal("ur:leaf/iejyihjkjygupyltla", urString);

        var ur2 = UR.FromUrString(urString);
        ur2.CheckType("leaf");
        var test2 = TestLeaf.FromUntaggedCbor(ur2.Cbor);
        Assert.Equal(test.S, test2.S);
    }
}

File: swift/BCUR/Tests/BCURTests/URCodableTests.swift
import Testing
import DCBOR
@testable import BCUR

private struct TestLeaf: URCodable, Equatable {
    static let cborTags: [BCTags.Tag] = [BCTags.Tag(24, "leaf")]

    let value: String

    init(_ value: String) {
        self.value = value
    }

    var untaggedCBOR: CBOR {
        value.cbor
    }

    init(untaggedCBOR: CBOR) throws {
        self.value = try String(cbor: untaggedCBOR)
    }
}

struct URCodableTests {
    @Test func testUrCodable() throws {
        let test = TestLeaf("test")
        let ur = test.ur()
        #expect(ur.urString == "ur:leaf/iejyihjkjygupyltla")

        let decoded = try TestLeaf.fromURString(ur.urString)
        #expect(decoded == test)
    }
}
